### PR TITLE
chore: add the stacks attribute

### PR DESCRIPTION
**Pull request by adomani**
This PR contains a subset of the code of #16191 and is probably not going to be merged.

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/stacks.20tags)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/StacksAttribute.lean [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/-
+Copyright (c) 2024 Dmaiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dmaiano Testa
```
Comment (bryangingechen) [inline Mathlib/Tactic/StacksAttribute.lean:4]:
```suggestion
Authors: Damiano Testa
```

Comment (adomani) [reply Mathlib/Tactic/StacksAttribute.lean]: :facepalm: thanks!

File: Mathlib/Tactic/StacksAttribute.lean [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+The `TAG` is mandatory.
+-/
+syntax (name := stacks) "stacks " (stackTags)? (str)? : attr
+
+initialize Lean.registerBuiltinAttribute {
+  name := `stacks
+  descr := "Apply a Stacks project tag to a theorem."
+  add := fun _decl stx _attrKind => Lean.withRef stx do
+    match stx with
+      | `(attr| stacks $_:stackTags $_:str) => return
```
Comment (mattrobball) [inline Mathlib/Tactic/StacksAttribute.lean:36]: Can we validate the `stackTag` here?

Comment (adomani) [reply Mathlib/Tactic/StacksAttribute.lean:41]: I just added a check that there are no spaces in the tag itself.  By "validate" you mean "make sure that such a tag exists"?  That should probably be done by a linter, right?

Comment (mattrobball) [reply Mathlib/Tactic/StacksAttribute.lean:41]: Sorry. What something like
```lean
syntax num | lit : stacksTagAtom
syntax stacksAtom stacksAtom stacksAtom stacksAtom stacksAtom : stacksTag
```
Is there some uppercase Latin character syntax category? Do we really have to write out `stacksAtom` 5 times?

Comment (adomani) [reply Mathlib/Tactic/StacksAttribute.lean:41]: Oh, the length 5 check can be done with the substring and also the uppercase check.

File: Mathlib/Tactic/StacksAttribute.lean [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+The `stacks` attribute.
+-/
+
+open Lean
+
+namespace Mathlib.Stacks
+
+/--
+The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
+This allows `044Q3` and `GH3F6` as possibilities.
```
Comment (bryangingechen) [inline Mathlib/Tactic/StacksAttribute.lean:18]: Maybe worth linking to the Stacks project page on tags: https://stacks.math.columbia.edu/tags

**Comment by mattrobball** [comment]
For the environment extension, do we want to store anything that map from the declaration name to the stacks tag as a string?

File: Mathlib/Tactic/StacksAttribute.lean [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+-/
+declare_syntax_cat stackTags
+
+@[inherit_doc Parser.Category.stackTags]
+syntax (num)? (ident)? : stackTags
```
Comment (mattrobball) [inline Mathlib/Tactic/StacksAttribute.lean:23]:
```suggestion
syntax (num)? (ident)? : stackTag
```

Comment (mattrobball) [reply Mathlib/Tactic/StacksAttribute.lean:23]: This is a single tag right?

File: Mathlib/Tactic/StacksAttribute.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Tactic/StacksAttribute.lean:50]: I have a slight preference for not using the double quotes, which a `str` would want.  If there is a way of parsing a string without quotes, then I would probably prefer that!

File: Mathlib/Tactic/StacksAttribute.lean (hunk 3 above)
Comment (adomani) [reply Mathlib/Tactic/StacksAttribute.lean:21]: Added, thanks!

**Comment by adomani** [comment]
> For the environment extension, do we want to store anything more than a map from the declaration name to the stacks tag as a string?

For the moment, I think probably not.  Maybe in the future we may want to add something else.

File: Mathlib/Tactic/StacksAttribute.lean [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Attributes
+
+/-!
+The `stacks` attribute.
+
+This allows tagging of mathlib lemmas with the corresponding
+[Tags](stacks.math.columbia.edu/tags) from the Stacks Project.
```
Comment (bryangingechen) [inline Mathlib/Tactic/StacksAttribute.lean:12]: I'm not sure if doc-gen4 is smart enough to link to the right place without https, but probably safest to include it anyways (also below):
```suggestion
[Tags](https://stacks.math.columbia.edu/tags) from the Stacks Project.
```

**Comment by adomani** [comment]
The "other" PR is getting merged, so I'll close this one.

Thank you all for your comments: they are part of the other PR anyway!

**Pull request closed without merging**